### PR TITLE
Fix: Additional null checks while provisioning team

### PR DESCRIPTION
**Pull request by gautamdsheth**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| New sample?      | no
| Related issues?  | fixes NA

#### What's in this Pull Request?

1) This PR adds additional null check while provisioning teams with additional settings like FunSettings, GuestSettings etc. 

2) Additionally, it removes compiler directives which are not needed as we already handle it "above".

3) If the Giphy content rating is set to Strict, that value is not respected, so fixed that issue with static class.

4) Added support for users with special characters in email address.

This PR fixes null issue when you try to provision a team from PnP template which is as below:

```
<pnp:Teams>    
    <pnp:Team DisplayName="Sample Team 01" Description="This is just a sample Team 01" Visibility="Public" Specialization="None">
      <pnp:Security>
        <pnp:Owners ClearExistingItems="true">
          <pnp:User UserPrincipalName="admin@tenant.onmicrosoft.com" />
          <pnp:User UserPrincipalName="meganb@tenant.onmicrosoft.com" />
        </pnp:Owners>
        <pnp:Members ClearExistingItems="true">
          <pnp:User UserPrincipalName="admin@tenant.onmicrosoft.com" />
          <pnp:User UserPrincipalName="meganb@tenant.onmicrosoft.com" />
        </pnp:Members>
      </pnp:Security>
    </pnp:Team>
</pnp:Teams>	
```

When i try to provision a Team using above template, the Team is not getting created, error is not thrown either but debugging the code, it gives a null reference error. This PR fixes it.

### CODE REVIEW

**Comment by jackpoz** [comment]
I also receive a null reference exception at https://github.com/SharePoint/PnP-Sites-Core/blob/master/Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs#L870 because PnPProvisioningContext.Current is null:
image

Does this PR handle this case too ? (I'm not even sure why it's null or how the rest of the provisioning engine handles the case of a null PnPProvisioningContext.Current)

**Comment by gautamdsheth** [comment]
@jackpoz  - good catch, it didn't handle that ! I have added that check now via a new commit. Thanks.

**Comment by jackpoz** [comment]
do you know how PnPProvisioningContext.Current is provided ? Is it the caller that has create a new PnPProvisioningContext or ?
Are you using Apply-PnPTenantTemplate or your own .NET app ? Could you try with Apply-PnPTenantTemplate too ?

**Comment by jackpoz** [comment]
After applying https://github.com/SharePoint/PnP-PowerShell/pull/2099 and this PR, Apply-PnPTenantTemplate can create a Team without throwing any exception 👍

**Comment by PaoloPia** [comment]
Hi,
Thanks for your contribution. About the PnPProvisioningContext object, here (https://www.youtube.com/watch?v=n6Mh9AzO9A8&t=3s) you can find a short video where I explain the purpose and how to use it. It is supposed to be provided by the caller in .NET code, and it is handled by PnP PowerShell automatically. About the suggested changes, I'm evaluating them. Thanks.

File: Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs [hunk 1]
```diff
@@ -594,7 +594,7 @@ private static bool SetGroupSecurity(PnPMonitoredScope scope, Team team, string
         /// <returns>Whether the Channels have been provisioned or not</returns>
         private static bool SetTeamChannels(PnPMonitoredScope scope, TokenParser parser, Team team, string teamId, string accessToken)
         {
-            if (team.Channels != null)
+            if (team.Channels.Count > 0)
```
Comment (PaoloPia) [inline Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs]: Hi, what is the purpose of this test? If the Channels collection is not null but still empty the foreach clause will be skipped. On the contrary if Channels is null and we test the Count property, we will get a null reference exception.

File: Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs [hunk 2]
```diff
@@ -622,66 +622,68 @@ private static bool SetTeamChannels(PnPMonitoredScope scope, TokenParser parser,
                     if (channelId == null) return false;
 
                     // If there are any Tabs for the current channel
-                    if (channel.Tabs == null || !channel.Tabs.Any()) continue;
-
-                    foreach (var tab in channel.Tabs)
+                    if (channel.Tabs.Count > 0)
```
Comment (PaoloPia) [inline Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs]: Same here, the foreach over an empty collection will simply skip it. I think we don't need to test Count > 0.

File: Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs [hunk 3]
```diff
@@ -622,66 +622,68 @@ private static bool SetTeamChannels(PnPMonitoredScope scope, TokenParser parser,
...
                     }
 
                     // TODO: Handle TabResources
                     // We need to define a "schema" for their settings
 
                     // If there are any messages for the current channel
-                    if (channel.Messages == null || !channel.Messages.Any()) continue;
-
-                    foreach (var message in channel.Messages)
+                    if (channel.Messages.Count > 0)
```
Comment (PaoloPia) [inline Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs]: And the same note applies here.

File: Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs [hunk 4]
```diff
@@ -622,66 +622,68 @@ private static bool SetTeamChannels(PnPMonitoredScope scope, TokenParser parser,
...
-                            accessToken,
-                            "NameAlreadyExists",
-                            CoreResources.Provisioning_ObjectHandlers_Teams_Team_TabAlreadyExists,
-                            "displayName",
-                            tab.DisplayName,
-                            CoreResources.Provisioning_ObjectHandlers_Teams_Team_ProvisioningError,
-                            canPatch: false);
-
-                        if (tabId == null) return false;
+                                tab.DisplayName,
```
Comment (PaoloPia) [inline Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/ObjectTeams.cs]: Could you please explain the purpose of this change?

**Comment by jackpoz** [comment]
> and it is handled by PnP PowerShell automatically

Erwin fixed that part in https://github.com/SharePoint/PnP-PowerShell/commit/e3b3a4e2470190af599229e33c794033cdc5f813 so that kind of explained why it didn't work before

**Comment by gautamdsheth** [comment]
closing this PR , had some merge conflicts :(

**Pull request closed without merging**